### ISSUE TITLE
SGX Enablement

### ISSUE THREAD

**Issue by rbradford** [cloud-hypervisor/cloud-hypervisor#1311]
Add support for exposing SGX to the guest.

**Comment by sboeuf** [comment]
I spent a bit of time looking at the proposed QEMU patches [here](https://github.com/intel/qemu-sgx/commits/master), and here are the main points we need to implement when enabling SGX for Cloud-Hypervisor:
- Based on the KVM host patches and the SGX driver, a new device `/dev/sgx/virt_epc` is going to be present on the host. This device is meant to be opened, which will provide a file descriptor that can be used to back a memory mapped region of _reserved_ memory.
- There is only one SGX EPC region that will be located after guest RAM and above 4G.
- The user can ask for multiple EPC sections, but they will all be contiguous in the global EPC region.
- The VMM will passthrough the SGX support based on the CPU host support. We might want to restrict some specific SGX features (by tweaking the CPUID leaves). We will never emulate some SGX features that are not part of the host CPU.
- The VMM will expose the global EPC region through e820 table as a _reserved_ region, and a new ACPI device will be added to the DSDT. This ACPI device will advertise start address and size of the EPC region, but it won't expose the number of EPC sections.
- The host kernel needs to be patched to include KVM support for SGX.
- The guest kernel needs to be patched to include the SGX driver, so the SGX device can be detected by the guest.

**Comment by sboeuf** [comment]
One important thing before starting implementing SGX for Cloud-Hypervisor:
- The VMM must create the EPC sections before the vCPUs are created. If this is not the case, the guest would not be able to access the EPC sections created after the vCPUs, because these sections are accessed through the CPUID leaves.
